docs2/site/docs/guides/migration3.md

```diff
@@ -0,0 +1,434 @@
+    public Task AfterValidationAsync(IExecutionContext context, IValidationResult validationResult)
+    {
+        var myContext = (MyContext)context.userContext;
+
+        // log validation error
+
+        return Task.CompletedTask;
+    }
+
+    ...
+}
+```
+
+### IResolveFieldContext and IResolveConnectionContext
+
+Field resolver methods now pass a reference to an `IResolveFieldContext` interface, rather than a
+`ResolveFieldContext` class. Inline lambda functions are typically unaffected, but if you define your
+resolvers separately, you will need to change the function signature.
+
+```csharp
+class MyGraphType : ObjectGraphType
+{
+    public MyGraphType()
+    {
+        Field("Name", resolve: x => "John Doe");
+        Field("Children", resolve: GetChildren);
+    }
+
+    //public IEnumerable<string> GetChildren(ResolveFieldContext context)
+    public IEnumerable<string> GetChildren(IResolveFieldContext context)
+    {
+        return new [] { "Jack", "Jill" };
+    }
+}
+```
+
+Also please note that all `IResolveFieldContext` and similar interfaces and classes have moved from
+the `GraphQL.Types` namespace to the `GraphQL` namespace. You may need to add a `using GraphQL;`
+statement to some of your files.
+
+### Connection Builders
+
+The connection builders have changed slightly. Please see https://graphql-dotnet.github.io/docs/getting-started/relay for current implementation details.
+
+### Field Middleware
+
+Field Middleware must be registered in the dependency injection container in order to be instantiated.
+You also need to implement the `IFieldMiddleware` interface on your custom middleware classes, and change
+the signature for the `Resolve` method to accept `IResolveFieldContext`.
+
+```csharp
+//class MyMiddleware
+class MyMiddleware : IFieldMiddleware
+{
+    //public async Task<object> Resolve(ResolveFieldContext context, FieldMiddlewareDelegate next)
+    public async Task<object> Resolve(IResolveFieldContext context, FieldMiddlewareDelegate next)
+    {
+        // your code here
+        var ret = await next(context);
+        // your code here
+        return ret;
+    }
+}
+```
+
+Please note that the delegate definition for `FieldMiddlewareDelegate` has changed as follows:
+
+```csharp
+// version 2.4.0
+public delegate Task<object> FieldMiddlewareDelegate(ResolveFieldContext context);
+// version 3.0
+public delegate Task<object> FieldMiddlewareDelegate(IResolveFieldContext context);
+```
+
+You also must ensure that your schema implements `IServiceProvider`. This is handled
+automatically if you inherit from `Schema`.
+
+See [Field Middleware](https://graphql-dotnet.github.io/docs/getting-started/field-middleware) for more details,
+including guidelines and restrictions on service lifetimes of middleware registered through your DI framework.
+
+### Data Loaders
+
+Data loaders now return an `IDataLoaderResult<T>` rather than a `Task<T>`. Field resolver signatures may need to change
```

Comment:
I've been going to review #1511 for quite some time. I couldn't do it because I had to review many of your other PRs (which in itself is not bad, you did a lot). Time is coming to the end of August. I'm not really sure if I can make a review in time. I would not want to release the 3.0 version (or what is the same to merge #1511 in master) with this feature, which has not been reviewed, despite the fact that it has been successfully applied by someone in production. Another reason is that I also have a set of changes that I would also like to add to the 3.0 version. There are backward incompatible things - middlewares redesign, changes in SDL printing and custom scalars. The dev teams in my company are also waiting for them (I am not yet talking about bug fixes which also needs to be done a lot). I would not want to release 4.0 immediately after the 3.0 version. Let's move the deadline for 3.0 to October 1, by a month. I will have a vacation almost all September and I will be able to devote much more time to this project. Recently, I simply could not spend all day thinking about a complex PR like #1511 one. So during next month, we will be able to prepare #1511 and other features as well. I believe it is better to work a little longer without rushing and produce a better product. Another month of delay isn't such a big deal if the previous version release was almost 2 years ago.

Reply:
I, respectfully, disagree.  There are multiple people who use #1511 already.  We are all here to maintain the project.  Let's just merge the dataloaders and keep going.  I've had this PR up since January.  I've done all the documentation.  I've made all these other changes we've been trying to get into 3.0.  Got a ton of stuff cleared off the issue list.  Anyone who wants to review my PR has had ample time to do so.  There's no reason to wait another month.

I understand releasing a few more features with 3.0, as long as we have a **SET** deadline and very good reason for the delay.  We can just merge the data loaders and release this as another beta.  If you have a dev team waiting for the features @sungam3r then good, let's get it done and release 3.0.  Other people have dev teams waiting for stuff, though, so we can't let it drag out indefinitely.  Pick a date and stick to it, regardless of the result.  We can always release another version.

If we do simply release another 3.0 beta (with data loaders) this weekend, then if there's any changes we need to make to the data loaders, they can be done in the interim.  Breaking changes are not a big problem.  Indefinite delays are.

Reply:
I fall into the camp of releasing 3.0 beta with the data loaders, 3.0, and releasing a 4.0 soon after if needed.  I can review #1511 tomorrow and propose to merge if I don't find anything major.

Reply:
Fine. I just wanted someone else to review.